packages/gatsby/src/cache-dir/component-renderer.js

```diff
@@ -55,22 +57,40 @@ class ComponentRenderer extends React.Component {
   // Check if the component or json have changed
   shouldComponentUpdate(nextProps, nextState) {
     if (
-      this.state.pageResources.component !== nextState.pageResources.component
+      this.state.pageResources.component !== nextState.pageResources.component ||
+      this.state.pageResources.json !== nextState.pageResources.json
     ) {
-      return true
-    }
-    if (this.state.pageResources.json !== nextState.pageResources.json) {
+      this.previousProps = this.props
+      this.previousPageResources = this.state.pageResources
       return true
     }
     return false
   }
 
   render() {
     if (this.state.pageResources) {
-      return createElement(this.state.pageResources.component, {
-        ...this.props,
-        ...this.state.pageResources.json,
+      const pluginResponses = apiRunner('componentRenderer', {
```

Comment:
Let's name this API `replacePageComponentRenderer` per https://www.gatsbyjs.org/docs/api-specification/#operators